guides/hack/02-source-code-fundamentals/37-namespaces.md

```diff
@@ -1,30 +1,39 @@
 
-A namespace may have *sub-namespaces*, where a sub-namespace name shares a common prefix with another namespace. For example, the
-namespace `Graphics` might have sub-namespaces `Graphics\TwoD` and `Graphics\ThreeD`, for two- and three-dimensional facilities,
-respectively. Apart from their common prefix, a namespace and its sub-namespaces have no special relationship. The namespace
-whose prefix is part of a sub-namespace need not actually exist for the sub-namespace to exist. That is, `NS1\Sub` can exist
-without `NS1`.
+The predefined constant `__NAMESPACE__` contains the name of the current namespace.
 
+## The Root Namespace
 In the absence of any namespace definition, the names of subsequent classes, interfaces, traits, functions, and constants are in
-the *default namespace*, which has no name, per se.
+the ***root namespace***, which is not named.
+
+## Sub-Namespaces
+A namespace can have ***sub-namespaces***, where a sub-namespace name shares a common prefix with another namespace. 
+
+For example, the namespace `Graphics` can have sub-namespaces `Graphics\TwoD` and `Graphics\ThreeD`, for two- and three-dimensional facilities,
+respectively. 
 
-The names of some types (such as `Exception`), constants, and library functions (such as `sqrt`) are defined outside any namespace.
-To refer unambiguously to such names, one can prefix them with a backslash (`\`), as in `\Exception` and `\sqrt`. The names of
-the standard types that are introduced with Hack belong to namespace `HH`.
+Apart from their common prefix, a namespace and its sub-namespaces have no special relationship. For example, `NS1\Sub` can exist without `NS1`. A named top-level namespace does not need to exist for a sub-namespace to exist.
 
+## Reserved Namespaces
 The namespaces `HH`, `PHP`, `php`, and sub-namespaces beginning with those prefixes are reserved for use by Hack.
 
-The pre-defined constant `__NAMESPACE__` contains the name of the current namespace.
+### The `HH` Namespace
+Built-in types like `vec`, `keyset`, `dict`, and `shape` exist in the `HH` namespace (e.g. `HH\Lib\Vec`).
 
-When the same namespace is defined in multiple scripts, and those scripts are combined into the same program, the namespace
-is considered the merger of its individual contributions.
+Other types, such as `Exception`, and constants, and library functions (such as `sqrt`) are in `HH` too. Prefix a backslash (`\`) to refer to these types or functions; for example: `\Exception`, `\sqrt`.
+
+### XHP and the `HTML` Namespace
+As of HHVM 4.73 and XHP-Lib v4, standard XHP elements like `<p>` are defined in `Facebook\XHP\HTML` (for this example, specifically `Facebook\XHP\p`).
 
-A namespace directive has two forms.  For example, Script1.php uses the semi-colon form:
+For more information, see [XHP Namespace Syntax](/hack/XHP/basic-usage#namespace-syntax).
+
+## Defining a Namespace
+Namespaces can be defined with semicolons (`;`) or brace delimiters (`{ ... }`).
```

Comment:
This feels a bit weird; how about:

namespace declarations can be file-scoped with `namespace MyNS;`, or block-scoped with `namespace MyNS { ... } ` ?

Reply:
Related: Do you think the word 'defined' is wrong in this context?

Reply:
The syntax is named `namespace_declaration` in [hhast/NamespaceDeclaration](https://github.com/hhvm/hhast/blob/main/codegen/syntax/NamespaceDeclaration.hack#L15). Definition does indeed sound a bit off, but I am biased by the internal names exposed via hhast.

Reply:
The reason I changed was to bring more unity: for example. if we say the phrase "define a function", I think we should also use the phrase "define a namespace", or change to "declare a function", etc.

Reply:
Indeed, I see `define` being used in [the function introduction](https://docs.hhvm.com/hack/functions/introduction) as well.

&lt;speculation&gt;
This might make a lot of sense to a person who works a lot with C/C++. In those languages a declaration and definition are notably different. You can give multiple declarations for the same function, but only one definition.
```CPP
/* declaration */
std::vector<uint_16> get_elements();
/* definition */
std::vector<uint_16> get_elements() {
  std::vector<int> elements = {1, 2, 3};
  return elements;
}
```
So if my_file.cpp defines the get_elements function, my_file.h would declare it. Other files may include the header file or repeat the declaration.
&lt;/speculation&gt;

In Hack, every definition of a function / classish / type alias / constant is also the only declaration. You can therefore use define and declare interchangeably in Hack.

Reply:
I think I'm going to keep it as declare in the context of namespaces and define in the context of functions.